### ISSUE TITLE
Subgroup branching bug

### ISSUE THREAD

**Issue by tomleavy** [mlswg/mls-protocol#623]
There is currently a logical flaw in the subgroup branching flow.

In order to branch, you need to compose a `Welcome` message . A Welcome message includes `EncryptedGroupSecrets` which is defined as follows: 

```
struct {
  KeyPackageRef new_member;
  HPKECiphertext encrypted_group_secrets;
} EncryptedGroupSecrets;
```

The problem with this is that the `new_member` field of type `KeyPackageRef` cannot be made for existing members of the group because the leaf nodes are now `LeafNode` types. `KeyPackageRef` is now only used to add new members to the group, so we can't reuse this for sub group initialization.

**Comment by kkohbrok** [comment]
I'm not sure I understand correctly. Are you suggesting when branching a group off of another group, one re-uses the leaves of the old group? If so, that's not how it's meant to work. The new group created in the process of subgroup branching should be independent from the "parent" group safe for the PSK, which all member have to provide upon creation of the group. In fact one should never use the same key material in the leaves of multiple groups except when the supply of key packages that a party provides via the DS is exhausted (see Section 17.4).

**Comment by tomleavy** [comment]
That was part of my question. When there were key packages in the leaves, you theoretically could fork the group from a given point in time and reuse the current tree, but I wasn't sure if that was a proper solution or not. Sounds like it isn't, but either way, I think there should be clarification in the document about this. Let me write up a quick PR on it.

**Comment by rohan-wire** [comment]
> I'm not sure I understand correctly. Are you suggesting when branching a group off of another group, one re-uses the leaves of the old group? If so, that's not how it's meant to work. The new group created in the process of subgroup branching should be independent from the "parent" group safe for the PSK, which all member have to provide upon creation of the group. In fact one should never use the same key material in the leaves of multiple groups except when the supply of key packages that a party provides via the DS is exhausted (see Section 17.4).

@kkohbrok Konrad, could you safely derive new keys for the members of the subgroup instead? If so, that would be an enormous increase in efficiency and (especially in the federated case) make group creation timing consistent enough to do dynamically.

**Comment by kkohbrok** [comment]
@rohan-wire Nice idea! But it's tricky, since we're dealing with public key pairs that the creator of the new group doesn't know the private keys for. If we were only using HPKE schemes that support updateable public key encryption, the creator of the new group could indeed update the key in the existing group and send the "delta" encrypted to each corresponding member (which would still represent an overhead, but we might be able to optimize this somehow). This is similar to the RTreeKEM mechanism that @psyoptix et al. proposed at some point. To go all the way here, we'd have to do the same for the public key material in the credentials as well, which is not something we had considered before. It would be a new primitive, something like an updateable signature scheme? Probably something to consider for HPKE/MLS 2.0.

We could also allow the re-use of LeadNodes for this particular case. It would be the simpler solution and I don't think it would be catastrophic, since we have to assume that LeafNode (or KeyPackage) key material is re-used in the "last resort" scenario anyway, but this would definitely be something we should discuss on the mailing list, as it would break a pretty basic assumption we've made until now.

**Comment by bifurcation** [comment]
Just catching up on this.  @tomleavy nice catch here.

It seems like we have two options here:
1. Require new KeyPackages (as in #624)
2. Invent some way to make a new group based on LeafNodes alone

I probably agree that we should do (1) in the short term, but (2) also has some appeal -- it doesn't feel right to require new KeyPackages for folks you're already in touch with -- despite it probably being more spec text.  So here's a sketch of (2) in the interest of completeness.

Suppose we defined a message of the following form:

```
struct {
  opaque new_group_id<V>;
  LeafNodeRef subgroup_members<V>;
  UpdatePath path;
  Extensions group_context_extensions<V>;
} Branch;
```

... which would pretty closely parallel to Commit.  You would send it in
MLSMessage and require a `confirmation_tag` in MLSMessageAuth.  Where Commit instructs the recipient to create a new epoch from the current one within the same group, Branch would instruct the recipient to create a new epoch that is the first epoch of a new group.

| Field           | Commit                               | Branch                                       |
|-----------------|--------------------------------------|----------------------------------------------|
| Group ID        | `group_id`                           | `new_group_id`                               |
| Epoch           | `epoch + 1`                          | `0`                                          |
| Tree            | Tree with proposals applied          | Tree with specified leaves, otherwise empty  |
| Transcript hash | Updated with Commit                  | `""`                                         |
| Extensions      | Carried over / specified in proposal | As specified                                 |
| PSKs            | Specified in proposal(s)             | Branch PSK from the current epoch            |

In both cases, you would form an epoch / GroupContext as specified, then use the `path` to ratchet into the new epoch.

I'm pretty sure that would work and have the expected security properties, though obviously formal analysis would be appreciated.  The main awkardness that occurs to me is that any members joining after the branch would have to provide new key packages.  And obviously, it would require a fair bit more spec text.

All that said, though, the exercise of working all that out has increased my impression that we should do (1) for now.  If we want something like (2) later, it can be done pretty cleanly in an extension.

**Comment by bifurcation** [comment]
Fixed by #624
